lib/middleware/authentication/authentication.test.js

```diff
@@ -0,0 +1,147 @@
+    expect(authentication).toBeInstanceOf(Function)
+  })
+
+  describe('how it runs in prod', () => {
+    beforeAll(() => {
+      process.env.NODE_ENV = 'production'
+      process.env.USE_AUTH = 'true'
+    })
+
+    describe('server with no username/password set', () => {
+      beforeAll(() => {
+        delete process.env.USERNAME
+        delete process.env.PASSWORD
+      })
+
+      beforeEach(() => {
+        console.error.mockClear()
+        authentication({}, res, next)
+      })
+
+      it('should return a console error', () => {
+        let consoleErrorMessage = console.error.mock.calls[0][0]
+        expect(consoleErrorMessage).toBe('Username or password is not set.')
+      })
+
+      it('should send marked up error message', () => {
+        let errorDisplayedToUser = res.send.mock.calls[0][0]
+        expect(errorDisplayedToUser).toBe('<h1>Error:</h1><p>Username or password not set. <a href="https://govuk-prototype-kit.herokuapp.com/docs/publishing-on-heroku#6-set-a-username-and-password">See guidance for setting these</a>.</p>')
+      })
+    })
+
+    describe('server with username/password set', () => {
+      beforeAll(() => {
+        process.env.USERNAME = 'secret-username'
+        process.env.PASSWORD = 'secure-password'
+      })
+
+      beforeEach(() => {
+        jest.clearAllMocks()
+        basicAuth.mockReturnValue(userDetails)
+        authentication({}, res, next)
+      })
+
+      it('should not return a console error', () => {
+        expect(console.error).not.toBeCalled()
+      })
+
+      it('should call basic-auth and return username/password', () => {
+        expect(basicAuth).toReturnWith(userDetails)
+      })
+
+      it('should not set authentication header', () => {
+        expect(res.set).not.toBeCalled()
+      })
+
+      it('should not return a status', () => {
+        expect(res.sendStatus).not.toBeCalled()
+      })
+
+      it('should progress to the next middleware', () => {
+        expect(next).toBeCalled()
+      })
+
+      describe('when a user supplies incorrect username/password', () => {
+        beforeEach(() => {
+          jest.clearAllMocks()
+          basicAuth.mockReturnValue(undefined)
+          authentication({}, res, next)
+        })
+
+        it('should call basic-auth and return username/password', () => {
+          expect(basicAuth).toReturnWith(undefined)
+        })
+
+        it('should set authentication header', () => {
+          expect(res.set).toHaveBeenCalledWith('WWW-Authenticate', 'Basic realm=Authorization Required')
+        })
+
+        it('should return 401/UnAuthorized', () => {
+          expect(res.sendStatus).toHaveBeenCalledWith(401)
+        })
+
+        it('should not progress to the next middleware', () => {
+          expect(next).not.toBeCalled()
+        })
+      })
+    })
+  })
+
+  describe('how it runs in non-production enviroment (dev by default)', () => {
+    beforeAll(() => {
+      delete process.env.NODE_ENV
```

Comment:
It would be good to have a comment here to explain why the environment variable is deleted.